### PR TITLE
Cube.SIZE Clean Up

### PR DESCRIPTION
**Pull request by Cyclonit**
I started cleaning up in preparation for using 32³ sized cubes, it is however a reasonable change by itself.

- Replaced most occurrences of the magic number 16 with Cube.SIZE in adequate places.
- Introduces constant Cube.SIZE_D=16.0 for decimal uses of the cubes' size.
- Replaced most occurrences of "& 15" and ">> 4" with blockToLocal and blockToCube where applicable. Some occurrences do not directly relate to cubes and where thus left unchanged.

### CODE REVIEW

File: src/main/java/cubicchunks/api/worldgen/populator/ICubicPopulator.java [hunk 1]
```diff
@@ -72,7 +73,7 @@
      */
     @Override
     default void generate(Random random, int chunkX, int chunkZ, World world, IChunkGenerator chunkGenerator, IChunkProvider chunkProvider) {
-        for (int y = 0; y <= 16; y++) {
+        for (int y = 0; y <= Cube.SIZE; y++) {
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/api/worldgen/populator/ICubicPopulator.java]: This 16 is unrelated to cube size. This is vanilla chunk height. Populator pregenerate those cubes to let default world generators do a job.

Comment (Cyclonit) [reply src/main/java/cubicchunks/api/worldgen/populator/ICubicPopulator.java]: ~~Will be fixed asap.~~

Fixed.

File: src/main/java/cubicchunks/asm/mixin/fixes/common/MixinTileEntityEndGateway.java [hunk 2]
```diff
@@ -69,9 +70,9 @@ private static int getChunkTopFilledSegmentFindSpawnpoint(Chunk chunk) {
      */
     @Overwrite
     private static Chunk getChunk(World world, Vec3d pos) {
-        Chunk chunk = world.getChunkFromChunkCoords(MathHelper.floor(pos.x / 16.0D), MathHelper.floor(pos.z / 16.0D));
+        Chunk chunk = world.getChunkFromChunkCoords(MathHelper.floor(pos.x / Cube.SIZE_D), MathHelper.floor(pos.z / Cube.SIZE_D));
         if (((IColumn) chunk).getCubicWorld().isCubicWorld()) {
-            for (int cubeY = 0; cubeY < 16; cubeY++) {
+            for (int cubeY = 0; cubeY < Cube.SIZE; cubeY++) {
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/asm/mixin/fixes/common/MixinTileEntityEndGateway.java]: Again, "16" here - vanilla chunk height in EBS-s.

File: src/main/java/cubicchunks/network/PacketCubeBlockChange.java [hunk 3]
```diff
@@ -68,14 +70,14 @@ public PacketCubeBlockChange(Cube cube, TShortCollection localAddresses) {
             int y = AddressTools.getLocalY(localAddress);
             int z = AddressTools.getLocalZ(localAddress);
             this.blockStates[i] = cube.getBlockState(x, y, z);
-            xzAddresses.add(x | z << 4);
+            xzAddresses.add(x | z << 4); // TODO: Is this the right way around?
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: This value ```x | z << 4``` is used in height map, therefore ```z``` value should be shifted to yours Cube.BIT_SIZE (5 for size = 32) or bits from "x" and "z" will overlap with each others.

Comment (Cyclonit) [reply src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: In a 16³ cube, both x and z are at most 4 bits in length. Thus the order in which they are packed doesn't matter in that regard or am I missing something?

File: src/main/java/cubicchunks/world/cube/Cube.java [hunk 4]
```diff
@@ -417,7 +419,7 @@ public void tickCubeServer(BooleanSupplier tryToTickFaster, Random rand) {
     public void randomTick(WorldServer worldServer, Random rand) {
         this.updateLCG = this.updateLCG * 3 + 1013904223;
         int j1 = updateLCG >> 2;
-        int localX = j1 & 15;
+        int localX = j1 & 15;           // TODO: Why not use AddressTools.getLocalX() instead? Uses the first 12 bits, instead of the first 4, 9 through 12 and 17 through 20.
         int localY = j1 >> 8 & 15;
         int localZ = j1 >> 16 & 15;
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/world/cube/Cube.java]: ```int localZ = blockToLocal(j1 >> 16)``` will work fine too. Unless you will decide to make Cube.size = 1024 or something.

File: src/main/java/cubicchunks/network/PacketCubeBlockChange.java (hunk 3 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: They are unpacket in a same <init> method at a line 80. Order matters here. Also note that height itself stuffed inside that value as well. So you would need to change it to ```v |= height << Cube.BIT_SIZE*2;```

Comment (Foghrye4) [reply src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: @Cyclonit also look how Client network handler handle ```int[] heightValues;``` field.

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: This should probably be changed to use a util method in class like AddressTools, together with all other similar things.

File: src/main/java/cubicchunks/world/cube/Cube.java (hunk 4 above)
Comment (Niko-sk2x) [reply src/main/java/cubicchunks/world/cube/Cube.java]: AddressTools probably can be used just fine here. The exact order of which bits are used for what doesn't really matter.

**Comment by Niko-sk2x** [comment]
Generally I think this is a step in the right direction by just looking at the description, even without 32x32x32 cubes (I didn't go over the code yet).

As for 32x32x32 cubes:

Changing to real 32x32x32 cubes probably isn't going to happen. Instead it would probably be better to just di the change for the most performance critical parts like PlayerCubeMap, which could load whole 32x32x32 sections, without any mention of 32^3 cubes anywhere else.

On-disk storage also could use 32x32x32 cubes, with a fallback 16x16x16 cube loader to do on the fly conversion (this may need some mesing around with RegionLib). NBT data would still be stored per 16x16x16 cubes, but 8 of them would be stored and compressed together in one region entry, just like vanilla does for 1x16x1 cubes.

The next place that could easily benefit from such change is mob spawning code. This would bring it much closer to vanilla performance for the same amount of blocks.

Random block ticks is another part that would benefit from it.

Other than these I don't see any places where changing to 32x32x32 cubes ould giv ay benefit. And the ones I listed can be changed without really introducing true 32x32x32 cubes.

File: src/main/java/cubicchunks/api/worldgen/populator/ICubicPopulator.java [hunk 5]
```diff
@@ -28,6 +28,7 @@
 import cubicchunks.api.worldgen.biome.CubicBiome;
 import cubicchunks.util.CubePos;
 import cubicchunks.world.ICubicWorld;
+import cubicchunks.world.cube.Cube;
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/api/worldgen/populator/ICubicPopulator.java]: Why?

File: src/main/java/cubicchunks/asm/mixin/core/client/MixinRenderGlobal.java [hunk 6]
```diff
@@ -170,9 +170,9 @@ public int getRenderChunkYPos(BlockPos pos) {
     @Overwrite
     private RenderChunk getRenderChunkOffset(BlockPos playerPos, RenderChunk renderChunkBase, EnumFacing facing) {
         BlockPos blockpos = renderChunkBase.getBlockPosOffset16(facing);
-        return MathHelper.abs(playerPos.getX() - blockpos.getX()) > this.renderDistanceChunks * 16 ? null :
-                MathHelper.abs(playerPos.getY() - blockpos.getY()) > this.renderDistanceChunks * 16 ? null :
-                        MathHelper.abs(playerPos.getZ() - blockpos.getZ()) > this.renderDistanceChunks * 16 ? null :
+        return MathHelper.abs(playerPos.getX() - blockpos.getX()) > this.renderDistanceChunks * Cube.SIZE ? null :
+                MathHelper.abs(playerPos.getY() - blockpos.getY()) > this.renderDistanceChunks * Cube.SIZE ? null :
+                        MathHelper.abs(playerPos.getZ() - blockpos.getZ()) > this.renderDistanceChunks * Cube.SIZE ? null :
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/core/client/MixinRenderGlobal.java]: While it removes a magic constant here, I think this part doesn't really allow for this value to be changed. Rendering will always stay 16x16x16 because anything bigger would add noticeable delays. So keep that in mind. Using 32x32x32 cubes is never going to be just changing a constant, you will need at least `VANILLA_SIZE=16` and `SIZE=32`, and using `VANILLA_SIZE` here

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinChunk_Cubes.java [hunk 7]
```diff
@@ -331,6 +331,7 @@ private boolean getBlockLightOpacity_isChunkLoadedCubeRedirect(Chunk chunk, int
     // TODO: Use @ModifyConstant with expandConditions when it's implemented
     @Overwrite
     public IBlockState getBlockState(final int x, final int y, final int z) {
+
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinChunk_Cubes.java]: Avoid formatting changes

File: src/main/java/cubicchunks/network/PacketCubeBlockChange.java (hunk 3 above)
Comment (Cyclonit) [reply src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: Wouldn't this just be AddressTools.getLocalAddress(int localX, int localZ)?

File: src/main/java/cubicchunks/worldgen/generator/custom/CustomTerrainGenerator.java [hunk 8]
```diff
@@ -76,8 +76,8 @@
 @MethodsReturnNonnullByDefault
 public class CustomTerrainGenerator extends BasicCubeGenerator {
 
-    private static final int CACHE_SIZE_2D = 16 * 16;
-    private static final int CACHE_SIZE_3D = 16 * 16 * 16;
+    private static final int CACHE_SIZE_2D = Cube.SIZE * Cube.SIZE;
+    private static final int CACHE_SIZE_3D = Cube.SIZE * Cube.SIZE * Cube.SIZE;
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/worldgen/generator/custom/CustomTerrainGenerator.java]: This 16 isn't directly related to cube size. The number is mostly a coincidence.

If you really wanted to use constants here, it would need to be something like 
```java
closestPowerOf2BiggerThan((Cube.SIZE / min(GENERATION_SECTION_SIZE_XZ, GENERATION_SECTION_SIZE_Y) + 1) * 3)
```
with `GENERATION_SECTION_SIZE_XZ=4` and `GENERATION_SECTION_SIZE_Y=8`

And the 5 below replaced with `(Cube.SIZE / min(GENERATION_SECTION_SIZE_XZ, GENERATION_SECTION_SIZE_Y) + 1)`

So this probably should be kept as is. Besides, the code will work fine anyway no matter the value, just slower or faster.

File: src/main/java/cubicchunks/worldgen/generator/custom/structure/CubicCaveGenerator.java [hunk 9]
```diff
@@ -66,7 +66,7 @@
      * Vanilla value: 7 Multiply by 16 and divide by 8: there are 16 cubes in a vanilla chunk, but only one cube per 8
      * has caves generated
      */
-    private static final int CAVE_RARITY = 16 * 7 / (2 * 2 * 2);
+    private static final int CAVE_RARITY = 16 * 7 / (2 * 2 * 2); // TODO: 32³ Cubes
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/worldgen/generator/custom/structure/CubicCaveGenerator.java]: Caves generation is already partially 32x32x32 and shouldn't be too hard to make work. But you would need whole 32x32x32 worldgen to go with it.

File: src/main/java/cubicchunks/worldgen/generator/custom/structure/feature/CubicFeatureGenerator.java [hunk 10]
```diff
@@ -115,7 +115,7 @@ public synchronized boolean generateStructure(World world, Random rand, CubePos
                     && structStart.getBoundingBox().intersectsWith(
                     new StructureBoundingBox(centerX, centerY, centerZ, centerX + 15, centerY + 15, centerZ + 15))) {
                 structStart.generateStructure(world, rand,
-                        new StructureBoundingBox(centerX, centerY, centerZ, centerX + 15, centerY + 15, centerZ + 15));
+                        new StructureBoundingBox(centerX, centerY, centerZ, centerX + 15, centerY + 15, centerZ + 15));   // TODO: Cube.SIZE - 1 instead of 15?
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/worldgen/generator/custom/structure/feature/CubicFeatureGenerator.java]: Yes, this should be changed to `Cube.SIZE - 1`

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]
Generally I don't think it will be very useful for making cube size 32x32x32, but good change anyway.

File: src/main/java/cubicchunks/network/PacketCubeBlockChange.java (hunk 3 above)
Comment (Niko-sk2x) [reply src/main/java/cubicchunks/network/PacketCubeBlockChange.java]: Probably, I haven't looked at AddressTools in a while.

**Comment by Cyclonit** [comment]
Of course, changing a simple variable won't convert CC to use 32³ cubes. But doing this first of all improves code maintainability in any case and if 32³ cubes ever where to happen, it would make things easier. But I think we're on the same page there.

I wonder if going for full 32³ cubes, while being more complicated at first, could be the better choice in the long run. Essentially, you propose to introduce a somewhat abstraction layer (which won't be as fast as a full 32³ implementation), for certain cases but not all, even though a 32³ implementation should be equally or better performing than the current 16³ implementation in almost all cases.

Secondly, using such an abstraction layer will most likely increase the overall complexity of the code even further and cause a lot of headaches in terms of maintainability in the future.

**Comment by Niko-sk2x** [comment]
We will need some abstraction layer either way. Vanilla expects 16xinfinityx16 columns, and this has to stay no matter what. We also don't really want to change ExtendedBlockStorage (while possible, it's very likely to break too much stuff and increase overall memory usage).

So if we want 16x16x16->32x32x32 abstraction layer, all we need is maybe some class that contains 2x2x2 cubes in a simple array. Easy access to arbitrary location. Used only in code we control, in ways we expect.

If we want 32x32x32->16x16x16 abstraction layer, then:
 * Either ExtendedBlockStorage is changed and we need a special code to handle accessing blocks through Chunk class, we still need mapping from ChunkXZ->Chunk, and if we also want to make 32x32x32 columns we need 2 Column classes. And it would make the already cmplicated Chunk mixin even more complicated. Ad Chunk may be used in ways we don't expect. And sponge compatibility will be a nightmare. I generally don't want to dela with changing EBS.
 * Or ExtendedBlockStorage stays as is, in which case all vanilla access would either go through 16x16x16->32x32x32 layer first, and then back through 16x16x16->32x32x32 for EBS access, or we would still keep the direct CubePos->16x16x16 cube mapping, in which case it's nearly equivalent to just changing some stuff and keeping other stuff.

So it's either something I really don't want to deal with or vanilla going through 2 additional abstraction layers, or basically equivalent to what I proposed.

**Review by Niko-sk2x — APPROVED** [review]
